### ISSUE TITLE
After several iteration, model.exec.output[0].asnumpy() get slower and slower !!

### ISSUE THREAD

**Issue by gramce** [apache/incubator-mxnet#2857]
My training code is in following, which is just like the code in the example.

```
def train(self, train_data, train_label,
          dev_data,
          dev_label,
          batch_size,
          optimizer='rmsprop',
          max_grad_norm=5.0,
          learning_rate=0.0005,
          epoch=200,
          eval_func=None):
    if eval_func is None:
        eval_func = lambda x, y: np.std(x.asnumpy() - y.asnumpy())

    opt = mx.optimizer.create(optimizer)
    opt.lr = learning_rate

    updater = mx.optimizer.get_updater(opt)

    for iteration in range(epoch):
        tic = time.time()

        for begin in range(0, train_data.shape[0], batch_size):
            batchX = train_data[begin:begin + batch_size]
            batchY = train_label[begin:begin + batch_size]

            if batchX.shape[0] != batch_size:
                continue

            print "data shape, ", self.data.shape
            print "label shape, ", self.label.shape
            print "input data shape, ", batchX.shape
            print "input label shape, ", batchY.shape
            self.data[:] = batchX
            self.label[:] = batchY

            self.mod_exec.forward(is_train=True)
            self.mod_exec.backward()

            train_error = eval_func(self.mod_exec.outputs[0], self.label)

            norm = 0.0
            for idx, weight, grad, name in self.param_blocks:
                grad /= batch_size
                l2_norm = mx.nd.norm(grad).asscalar()
                norm += l2_norm * l2_norm

            norm = math.sqrt(norm)
            for idx, weight, grad, name in self.param_blocks:
                if norm > max_grad_norm:
                    grad *= (max_grad_norm / norm)
                updater(idx, grad, weight)
                grad[:] = 0.0

            if dev_data is not None and dev_label is not None:
                for begin in range(0, dev_data.shape[0], batch_size):
                    batchX = dev_data[begin:begin + batch_size]
                    batchY = dev_label[begin:begin + batch_size]

                    if batchX.shape[0] != batch_size:
                        continue

                    self.data = batchX
                    self.label = batchY
                    self.mod_exec.forward(is_train=False)

                    dev_error = eval_func(self.mod_exec.outputs[0], self.label)

            if iteration % 50 == 0 and iteration > 0:
                opt.lr *= 0.5
                logging.info("reset optimizer learning rate to {}".format(opt.lr))

            toc = time.time()
            train_time = toc - tic
            logging.info("Iter {} Train: Time:{:.3f}s, Training Error:{:.3f} Dev Error:{:.3f}".format(iteration,
                                                                                                      train_time,
                                                                                                      train_error,
                                                                                                      dev_error))
```

**Comment by piiswrong** [comment]
How much slower? Your GPU is heating up and slowing down automatically. So some performance drop is normal

**Comment by gramce** [comment]
@piiswrong 
I don't think it is the problem. I test my work on cpu. In the first several iterations, the asnumpy cost about 6s. However, after 50 iterations, I waited for about twenty minutes. And the time cost of "forward" "backward" "update(weight)" is not increasing significantly.

And there's another problem now.  I set the num of iterations equals 1. After the iteration finished, the program can't quit. And the memory seemed not be released and continuously  occupied until it is eat up.

**Comment by gramce** [comment]
@piiswrong 

After several experiments, we found that as the iterations number increase, the time costs of program quitting increase significant. Increase the number of parameters will also increase the time cost.

**Comment by piiswrong** [comment]
Post your log and I'll likely be able to tell if it's normal or not

**Comment by gramce** [comment]
@piiswrong 
 it convenient to send you my code？
It could be found here.
https://github.com/gramce/Demo/blob/master/demo.py

**Comment by sxjscience** [comment]
@gramce You can try to simplify the code first (like using dummy data for training). It would be much easier for us to find the cause if we have a minimal reproducible example.

**Comment by piiswrong** [comment]
I need to see your log, not code, as the problem is likely specific to your machine

**Comment by gramce** [comment]
@piiswrong 

I am sorry about that I don't know where to find the log. Is there anyway that you could tell me what log you need? And I'll add it to my code.

However, I have tried my code on at least five machines and it seems nothing gets better.

Thanks.

**Comment by gramce** [comment]
@sxjscience 

OK. I'll try to simplify my code. Thanks.

**Comment by piiswrong** [comment]
How do you know it's getting slower? Show me the speed over time

**Comment by gramce** [comment]
@piiswrong 

I print the "train_accuracy" every 50 iterations. However, I found the program always hang when it try to get the model.exec.output[0] and can't calculate the "training accuracy" at all. So i tried to print the time cost every iterations. The log shows at least at the first several iterations, the "training accuracy" could be calculate correctly.

**Comment by gramce** [comment]
@piiswrong 

Here's part of the first several iterations' log.

iter_0batch_0forward_cost=5.00679016113e-05
iter_0batch_0backward_cost=5.50746917725e-05
iter_0batch_0eval_time=6.83129715919
iter_0batch_0update_cost=0.0371310710907
iter_0batch_0total_train_cost=6.86866593361
iter_0batch_1forward_cost=4.29153442383e-05
iter_0batch_1backward_cost=4.6968460083e-05
iter_0batch_1eval_time=25.0118870735
iter_0batch_1update_cost=0.0289700031281
iter_0batch_1total_train_cost=25.0410130024
iter_0batch_2forward_cost=3.81469726562e-05
iter_0batch_2backward_cost=4.00543212891e-05
iter_0batch_2eval_time=21.8669610023
iter_0batch_2update_cost=0.0204961299896
iter_0batch_2total_train_cost=21.8876068592
iter_0batch_3forward_cost=3.79085540771e-05
iter_0batch_3backward_cost=3.69548797607e-05
iter_0batch_3eval_time=21.7102849483
iter_0batch_3update_cost=0.0256681442261
iter_0batch_3total_train_cost=21.736096859
iter_1batch_0forward_cost=3.60012054443e-05
iter_1batch_0backward_cost=3.69548797607e-05
iter_1batch_0eval_time=21.6094560623
iter_1batch_0update_cost=0.0121791362762
iter_1batch_0total_train_cost=21.6217770576
iter_1batch_1forward_cost=3.60012054443e-05
iter_1batch_1backward_cost=3.60012054443e-05
iter_1batch_1eval_time=22.7317869663
iter_1batch_1update_cost=0.0251770019531
iter_1batch_1total_train_cost=22.7571120262
iter_1batch_2forward_cost=3.69548797607e-05
iter_1batch_2backward_cost=3.79085540771e-05

**Comment by gramce** [comment]
Here is what we are going to do.

**Comment by starimpact** [comment]
maybe size of the data transferred from gpu to cpu is bigger and bigger.

**Comment by back2yes** [comment]
...Unfortunately, I met the same problem with you.....
But I don't know how to solve it..

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
